src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java

```diff
@@ -0,0 +1,250 @@
+    }
+
+    @Override
+    protected boolean isJobCanceled() {
+        return craftingTask != null && craftingTask.percentageCompleted() < 100 && autocraftingComponent.getStatuses().stream().noneMatch(task -> task.info().id() == craftingTask.info().id());
+    }
+
+    @Override
+    public Object getParsedRequestedItemImpl() {
+        return RsApi.getObjectFromResourceKey(toCraft, amount);
+    }
+
+    @Override
+    public long getElapsedTimeImpl() {
+        if (craftingTask == null) {
+            return -1;
+        }
+        return System.currentTimeMillis() - craftingTask.info().startTime();
+    }
+
+    @Override
+    public long getTotalItemsImpl() {
+        if (craftingTask == null) {
+            return -1;
+        }
+
+        return craftingTask.items().size();
+    }
+
+    @Override
+    public long getItemProgressImpl() {
+        return craftingTask == null ? 0 : craftingTask.items().stream().mapToLong(TaskStatus.Item::stored).sum();
+    }
+
+    @Override
+    public Object getEmittedItemsImpl() {
+        if (preview == null) {
+            return Collections.emptyList();
+        }
+
+        return preview.items().stream().filter(item -> item.toCraft() > 0).map(item -> RsApi.getObjectFromResourceKey(item.resource(), item.toCraft())).collect(Collectors.toList());
+    }
+
+    @Override
+    public Object getUsedItemsImpl() {
+        // Not supported for rs2
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Object getMissingItemsImpl() {
+        if (preview == null) {
+            return Collections.emptyList();
+        }
+
+        return preview.items().stream().filter(item -> item.missing() > 0).map(item -> RsApi.getObjectFromResourceKey(item.resource(), item.missing())).collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean hasMultiplePathsImpl() {
+        // Not supported for rs2
+        return false;
+    }
+
+    @Override
+    public Object getFinalOutputImpl() {
+        if (preview == null) {
+            return Collections.emptyList();
+        }
+
+        return preview.outputsOfPatternWithCycle().stream().map(RsApi::getObjectFromResourceAmount).collect(Collectors.toList());
+    }
+
+    public TaskStatus getCraftingTask() {
+        return craftingTask;
+    }
+
+    @Override
+    public boolean cancelImpl() {
+        if (isJobDone() || isJobCanceled()) {
+            return false;
+        }
+        autocraftingComponent.cancel(craftingTask.info().id());
+        return true;
+    }
+
+    @Override
+    public void tick() {
+        super.tick();
+        // RS2 does re-create a new instance of the TaskStatus when something changes - so we actually need to re fetch that from the system
+        // It's cursed, but the only way currently
+        if (craftingTask != null) {
+            for (TaskStatus status : autocraftingComponent.getStatuses()) {
+                if (status.info().id() == craftingTask.info().id()) {
```

Comment:
are you sure `==` always works here? or do you think `equals` can be safer?

Reply:
I am 99% sure that it always uses the same instance
But to be safe, I changed it to equals